### PR TITLE
search: clients use 'standard' query interpretation

### PR DESCRIPTION
**Pull request by rvantonder**
Stacked on https://github.com/sourcegraph/sourcegraph/pull/37894.

This switches clients over to use 'standard' query interpretation, which corresponds to the default interpretation in the new V3 version for search. From docstring:

> History:
> - V3 - default to standard interpretation: Interpret patterns literally, or regular expression when enclosed by /.../ ([RFC 675](https://docs.google.com/document/d/1KebTihO_jrPcLblgeT4X428kfgs42NkFz_3kCr8bT0M/edit)).
> - V2 - default to interpreting patterns literally only.
> - V1 - default to interpreting patterns as regular expressions.
> - None - Anything before 3.9.0 will not pass a version parameter and defaults to V1.

The _mechanism_ for making this switch in the client is to rename (almost) all the instances of `SearchPatternType.literal` to `SearchPatternType.standard` in the client. It is the "correct" mechanism to use, since it is effectively source of truth for pattern interpretation because it overrides any version, and pervades all search requests in the client. The ideal state, though, would be to remove the existence of `patterntype` if/when we can decide there is only ever one query string interpretation. While 'standard' interpretation gets us closer to that goal, it's not a near term goal. The way it gets us closer to this goal is:

- The change effectively deprecates `patterntype:literal`
- The change makes running regular expressions easy, and so hopefully subsumes the need for `patterntype:regexp`. To be confident about this change subsuming `patterntype:regexp` will take some time to propose and evaluate, and also more changes to remove.

About `LATEST_VERSION`, this is updated to `V3` everywhere. But version doesn't effectively "do anything" for _our_ client today (or even client <-> server) interaction, because all those instances get the explicit `patterntype` (now `standard`), which overrides the version. But, in a world where we remove `patterntype`, the fallback to use version will indeed be useful and correct. Right now, the only utility of version is if some outside user/client developed against Sourcegraph some app that uses V1 or V2 and they _don't_ set the `patterntype`.

### Note these behavioral changes with query interactions in the client!

**Any new queries created will use `patterntype:standard` by default unless specified otherwise**. This applies to new searches in the client that do not explicitly specify `patterntype`. This affects at least:
  - new searches in the recent searches homepage (@sourcegraph/search-product)
  - code monitors, though it needs a small corresponding backend change to honor version (@sourcegraph/search-product)
  - notebooks (@sourcegraph/search-product)
  - _Maybe_ code insights, please check the affected code in this PR. I'm not clued up entirely about insights creation, but I seem to recall  you make users explicitly type out `patterntype:regexp` (@sourcegraph/code-insights)
  - Don't know if/how it affects @sourcegraph/batchers, depends on how the client calls the backend. I suggest updating to `patterntype:standard` and `V3`.

**Any instances that set `search.defaultPatternType` in settings is _unaffected_ by this change, since it overrides the default interpretation**.

**Previously saved queries preserve `patterntype:literal`**. This is true for existing saved queries displayed in the users homepage, codemonitors, and any other client that saves `patterntype` values in the backend and displays them in the client. 

 - It's basically impossible to try do any sort of migration here, because maybe it breaks user expectations. The only reasonable/maintainable thing is to move forward as of this release with the new behavior.

 - This means if a user clicks a link that does `patterntype:literal`, they are going to stay in the webapp with `patterntype:literal` unless they:
   - explicitly add `patterntype:standard` in the query; OR
   - remove `patterntype` from the URL; OR
   - toggle regex on/off (toggling will throw them into `patterntype:standard` after being toggled off)
    
     - Aside: it would be nice if the user could clear their recent searches to avoid "tricking" themselves into `patterntype:literal` saved queries, but that's not something I'm putting on my plate. Users have request to delete saved queries irrespective, so this is a good thing to add at some point.

I will deal with `src-cli` separately.

## Test plan

Test run against this change. Manual testing. Still struggling to get `main-dry-run` to build the server, don't know why!

### CODE REVIEW

File: client/web/src/enterprise/code-monitoring/components/FormTriggerArea.tsx [hunk 1]
```diff
@@ -250,7 +250,7 @@ export const FormTriggerArea: React.FunctionComponent<React.PropsWithChildren<Tr
                                 <Link
                                     to={`/search?${buildSearchURLQuery(
                                         queryState.query,
-                                        SearchPatternType.literal,
+                                        SearchPatternType.standard,
                                         false
                                     )}`}
                                     target="_blank"
```
Comment (limitedmage) [inline client/web/src/enterprise/code-monitoring/components/FormTriggerArea.tsx:256]: The first `ValidQueryChecklistItem` below probably needs its copy modified in some way. Maybe something like "must not be `patternType:structural`"? (Looking at the code above I don't think we even validate this correctly, might be a bug.)

**Review by limitedmage — COMMENTED** [review]
Early review of this looks great. Some questions on follow-up work:

- Should the regex toggle in the search box be removed after this work is done? Seems irrelevant now. (If a search is opened that has `patternType:regex`, it should be shown in the search box, similar to how contexts work.)
- Should we remove the patternType requirement in saved searches?
- How does lucky/auto search fit into this?
- Will filters still require regex or can we make them support globbing at some point (eg. use globbing by default but regex when wrapped in `//`)

File: CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Changed
 
+- **Search: Queries with patterns surrounded by** `/.../` **will now be interpreted at regular expressions.** Existing recent searches (e.g., links in your home page) or code monitors are unaffected. If you happen to use an existing older link and want `/.../` to work as a regular expression, add `patterntype:standard` to the query. New queries and code monitors will interpret `/.../` as regular expressions. [#38141](https://github.com/sourcegraph/sourcegraph/pull/38141).
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply CHANGELOG.md:24]: 🙏

File: client/shared/src/search/stream.ts [hunk 3]
```diff
@@ -10,9 +10,13 @@ import { SearchPatternType } from '../graphql-operations'
 import { SymbolKind } from '../schema'
 
 // The latest supported version of our search syntax. Users should never be able to determine the search version.
-// The version is set based on the release tag of the instance. Anything before 3.9.0 will not pass a version parameter,
-// and will therefore default to V1.
-export const LATEST_VERSION = 'V2'
+// The version is set based on the release tag of the instance.
+// History:
+// V3 - default to standard interpretation (RFC 675): Interpret patterns enclosed by /.../ as regular expressions. Interpret patterns literally otherwise.
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply client/shared/src/search/stream.ts:15]: Yes. No changes to filters.

**Comment by rvantonder** [comment]
@limitedmage nice questions. You're a bit early (CODENOTIFY is eager haha) but this change is basically ready to ship.

> Should the regex toggle in the search box be removed after this work is done? Seems irrelevant now. (If a search is opened that has patternType:regex, it should be shown in the search box, similar to how contexts work.)

Gonna quote myself here:

> The change makes running regular expressions easy, and so hopefully subsumes the need for `patterntype:regexp`. To be confident about this change subsuming `patterntype:regexp` will take some time to propose and evaluate, and also more changes to remove.

Basically, I want to give it some time where it co-exists, because it's not trivial to just remove. If a search is opened with `patterntype:regexp` (e.g., URL) and the toggles did not exist, then perhaps it would wrap all patterns in `/.../` to preserve backwards compatibility. This is a layer that I have low appetite to tackle and we can revisit once there's enough consensus that the button is redundant. It's OK for this "sort of" redundant button thing to co-exist right now because most users are not going to click that they can use `/.../` in the default search bar yet. Leaving it some time might help us more gracefully remove the toggle in time.

> Should we remove the `patternType` requirement in saved searches?

The person who owns saved searches can make that call. It would be a favorable thing, IMO, to remove the need for `patterntype` in saved searches. Basically, validate that if the user puts "patterntype" tell them "nope! but you can use `/.../` for regular expressions now".

> How does lucky/auto search fit into this?

It fits seamlessly. In fact lucky search can suggest "oh did you mean regular expression" by wrapping a pattern such as `foo.*bar` in `/foo.*bar/` without needing to suggest `patterntype:regexp` or switching the "mode". This is one of the main reasons I'm accelerating "standard" interpretation, it plays much better with suggestions.

> Will filters still require regex or can we make them support globbing at some point (eg. use globbing by default but regex when wrapped in //)

This is related but orthogonal effort to this PR/change. There is discussion on `filter:/xyz/` syntax in [RFC 675](https://docs.google.com/document/d/1KebTihO_jrPcLblgeT4X428kfgs42NkFz_3kCr8bT0M/edit#). The complexity for this change is too big of an ask for me to undertake right now (read: have to think harder about safely doing this).

File: client/web/src/enterprise/code-monitoring/components/FormTriggerArea.tsx (hunk 1 above)
Comment (rvantonder) [reply client/web/src/enterprise/code-monitoring/components/FormTriggerArea.tsx:256]: I'd recommend maybe a separate PR where code owners here update as appropriate. My read on this is: this change doesn't ~really break the validity of the check (at least, not enough that I want to insert that change into this PR), but it does give an opportunity to remove `patterntype` and talking about it. It seems to me that changing:

> hint="Code monitors support literal and regex search. Searches are literal by default."

to

> hint="Code monitors search patterns literally by default. Use `/regexp/` syntax to use search with expressions."

could work. Or even remove entirely. Kinda your call.

**Comment by coury-clark** [comment]
Sorry if this is obvious but is there a way to escape `/.../` in standard mode, or is the only option there to use literal search mode?

**Comment by rvantonder** [comment]
> Sorry if this is obvious but is there a way to escape /.../ in standard mode, or is the only option there to use literal search mode?

- The preferred recommendation is: `content:"/hi/"`

You can also use the syntax itself, if you prefer, since regex has it has it's own escape character `\` in the language:

- `/\/hi\//`

We do not recognize `\` as an escape sequence in _our_ language in the standard mode (and neither do other search engines I've interacted with). We will keep interpreting quotes literally.

**Review by camdencheek — APPROVED** [review]

File: client/web/src/enterprise/insights/core/backend/utils/create-line-chart-content.ts [hunk 4]
```diff
@@ -104,7 +104,7 @@ export function generateLinkURL(input: GenerateLinkInput): string {
     const beforeFilter = `before:${before}`
     const dateFilters = `${afterFilter} ${beforeFilter}`
     const diffQuery = `${contextFilter} ${scopeRepoFilters} ${repoFilter} type:diff ${dateFilters} ${query}`
-    const searchQueryParameter = buildSearchURLQuery(diffQuery, SearchPatternType.literal, false)
+    const searchQueryParameter = buildSearchURLQuery(diffQuery, SearchPatternType.standard, false)
```
Comment (rvantonder) [inline client/web/src/enterprise/insights/core/backend/utils/create-line-chart-content.ts:107]: @sourcegraph/code-insights can you please audit whether this change is reasonable to you, or would you like to keep `SearchPatternType.literal` here?

File: client/web/src/enterprise/search/stats/SearchStatsLanguages.tsx [hunk 5]
```diff
@@ -55,7 +55,7 @@ export const SearchStatsLanguages: React.FunctionComponent<React.PropsWithChildr
 
     const urlToSearchWithExtraQuery = useCallback(
         (extraQuery: string) =>
-            `/search?${buildSearchURLQuery(`${query} ${extraQuery}`, SearchPatternType.literal, false)}`,
+            `/search?${buildSearchURLQuery(`${query} ${extraQuery}`, SearchPatternType.standard, false)}`,
```
Comment (rvantonder) [inline client/web/src/enterprise/search/stats/SearchStatsLanguages.tsx:58]: @sourcegraph/code-insights can you please audit whether this change is reasonable to you, or would you like to keep `SearchPatternType.literal` here?

File: client/web/src/enterprise/insights/core/backend/utils/create-line-chart-content.ts (hunk 4 above)
Comment (coury-clark) [reply client/web/src/enterprise/insights/core/backend/utils/create-line-chart-content.ts:107]: Let's keep it literal

File: client/web/src/enterprise/search/stats/SearchStatsLanguages.tsx (hunk 5 above)
Comment (coury-clark) [reply client/web/src/enterprise/search/stats/SearchStatsLanguages.tsx:58]: Let's keep it literal

**Comment by rvantonder** [comment]
Time to merge this. Will be monitoring things closely to see if anything unexpected ends up breaking.

File: client/shared/src/util/url.ts [hunk 6]
```diff
@@ -516,6 +519,41 @@ export function withWorkspaceRootInputRevision(
...
+                      range: { start: 0, end: 0 },
+                      field: createLiteral('content', { start: 0, end: 0 }, false),
+                      value: createLiteral(`/${token.value}/`, { start: 0, end: 0 }, true),
+                      negated: false /** FIXME */,
+                  }
+                : token
+        )
+    )
+
+    console.log('mapped to standard')
```
Comment (valerybugakov) [inline client/shared/src/util/url.ts:549]: Hey @rvantonder, could you remove `console.log()` calls from the production bundle?

File: client/shared/src/util/url.ts [hunk 7]
```diff
@@ -525,13 +563,20 @@ export function withWorkspaceRootInputRevision(
...
-    patternType: SearchPatternType,
+    queryInput: string,
+    patternTypeInput: SearchPatternType,
     caseSensitive: boolean,
     searchContextSpec?: string,
     searchParametersList?: { key: string; value: string }[]
 ): string {
     const searchParameters = new URLSearchParams()
+
+    console.log(`patterntype is ${patternTypeInput}`)
```
Comment (valerybugakov) [inline client/shared/src/util/url.ts:574]: Hey @rvantonder, could you remove `console.log()` calls from the production bundle?

**Review by valerybugakov — COMMENTED** [review]
The Frontend Platform team works on the RFC for the client application logger that would be useful for debugging purposes and will be automatically disabled in the production environment. In the meantime, let's keep the console output clean to keep the signal/noise ratio high.

File: client/shared/src/util/url.ts (hunk 6 above)
Comment (rvantonder) [reply client/shared/src/util/url.ts:549]: ugh, sorry!

Comment (rvantonder) [reply client/shared/src/util/url.ts:549]: (this needs a lint 😆 )

Comment (rvantonder) [reply client/shared/src/util/url.ts:549]: (the console.logs were not intended to be merged)

Comment (rvantonder) [reply client/shared/src/util/url.ts:549]: turns out this file shouldn't have been changed and I should have unstaged before pushing 😖

Comment (valerybugakov) [reply client/shared/src/util/url.ts:549]: > this needs a lint 😆

Going to try enabling it today 👍

Comment (rvantonder) [reply client/shared/src/util/url.ts:549]: Future me thanks you 🧡

**Pull request merged**